qiskit/test/_canonical.py

```diff
@@ -0,0 +1,123 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 20022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility methods for canonicalising various Qiskit objects, to help with testing."""
+
+import threading
+
+from qiskit.circuit import (
+    BreakLoopOp,
+    CircuitInstruction,
+    ContinueLoopOp,
+    ControlFlowOp,
+    ForLoopOp,
+    Parameter,
+    QuantumCircuit,
+)
+
+
+class _CanonicalParametersIterator:
+    """An object that, when iterated through, will produce the same sequence of parameters as every
+    other instance of this iterator."""
+
+    __parameters = []
+    __mutex = threading.Lock()
+
+    def __init__(self):
+        self._counter = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        with self.__mutex:
```

Comment:
I wasn't super 100% sure how stestr handled things, and thought there was a risk there might be some threading somewhere.  Since the `__next__` operation isn't atomic, I just threw this in to be on the safe side.

Reply:
Ah, yeah stestr's parallelism is much coarser grained. It just uses subprocess to manually launch a new python interpreter for each test worker. Global state is isolated between worker processes so you don't have to worry about a race between test workers modifying shared state here.